106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+
+Here is an example of what the configured `Kafka` custom resource could look like:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+  # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the skew rebalance mode
+      - mode: skew
+        template:
+          name: my-skew-rebalance-template
+```
+
+It is also possible to use the default Cruise Control rebalancing configuration by omitting the `template` field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the default Cruise Control rebalancing configuration 
+      - mode: skew
+```
+
+The auto-rebalance configuration for the `spec.cruiseControl.autoRebalance.template` property in the `Kafka` custom resource is provided through a `KafkaRebalance` custom resource defined as a "template".
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance-template: true` annotation set.
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
+The user can specify rebalancing goals and other configuration for rebalancing, within the resource.
+
+Here is an example template:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-skew-rebalance-template
+  annotations:
+    strimzi.io/rebalance-template: "true" # specifies that this KafkaRebalance is a rebalance configuration template
+spec:
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+When the "template" is set the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource based on the "template" when an anomaly is detected and notified by the `AnomalyDetectorNotifier`
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one and also adds the `spec.mode` to it.
+The generated `KafkaRebalance` custom resource will be called `<my-cluster-name>-auto-rebalancing-skew-anomalyId`, where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, and `skew` refers to applying the rebalance to all the brokers and the anomalyId would be retrieved from the notifier.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-cluster-auto-rebalancing-add-brokers
+  finalizers:
+    - strimzi.io/auto-rebalancing
+spec:
+  mode: skew
```

Comment:
We are using "full" here which means a rebalancing which takes into account all brokers in the cluster (vs add-brokers and remove-brokers which take into account you are adding/removing specific brokers).

Reply:
So it will be always `full` and there will be no new mode like `imbalance` or `skew`, right?

Reply:
the `imbalance` term is the higher level definition of the mode for the "auto-rebalancing" alongside `add-brokers`, `remove-brokers`. But when a `KafkaRebalance` custom resource is created, the mode will be `full` which within the Strimzi rebalancing operator means calling the Cruise Control API to run a rebalancing taking all brokers into account.
To make it clearer, we could have done the same in the past by having the autorebalancing mode definitions of `scalingup` and `scalingdown` then mapped to the `add-brokers` and `remove-brokers` at `KafkaRebalance` level but we went using the same naming instead.